Refactor `BusinessATStep` tests to @testing-library/react

brookewp opened the pull request:
#### Proposed Changes

* This PR refactors the `BusinessATStep` component to use `@testing-library/react` instead of `enzyme`.

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Verify tests still pass: `yarn test-client client/components/marketing-survey/cancel-purchase-form/step-components/test/business-at-step.jsx`

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #63409

brookewp committed:
migrate tests from enyzme to testing-library

tyxla commented on client/components/marketing-survey/cancel-purchase-form/step-components/test/business-at-step.jsx:
```diff
@@ -1,75 +1,84 @@
-			wrapper = shallow( <BusinessATStep recordTracksEvent={ noop } translate={ translate } /> );
-		} );
+describe( 'rendering translated content', () => {
+	const translate = ( content ) => `Translated: ${ content }`;
```
Prefixing the content with `Translated:` is a clever hack to make sure we've actually ran `translate()`. Alternatively, we could also assert that the mock has been called: `expect( translate ).toHaveBeenCalled()`.

tyxla commented on client/components/marketing-survey/cancel-purchase-form/step-components/test/business-at-step.jsx:
```diff
@@ -1,75 +1,84 @@
-			);
-		} );
+		link.addEventListener( 'click', ( event ) => event.preventDefault(), false );
+		await userEvent.click( link );
```
One thing that we haven't been doing so far and we should, is calling `userEvent.setup()` before actually firing any events. See https://testing-library.com/docs/user-event/setup/ for more info.

tyxla commented on client/components/marketing-survey/cancel-purchase-form/step-components/test/business-at-step.jsx:
```diff
@@ -1,75 +1,84 @@
-		} );
+	test( 'should fire tracks event for theme support link when clicked', async () => {
+		render( <BusinessATStep translate={ translate } recordTracksEvent={ recordTracksEvent } /> );
+		const link = screen.queryAllByRole( 'link' )[ 1 ];
```
We could as well use `data-testid` in mocks to be even more specific if you don't want to traverse throughout all the links found.

brookewp replied:
Hmm I'm not sure about how it would be more specific in this case based on my understanding of `data-testid` from the docs: 

> In the spirit of [the guiding principles](https://testing-library.com/docs/guiding-principles), it is recommended to use this only after the other queries don't work for your use case. Using data-testid attributes do not resemble how your software is used and should be avoided if possible.

In this case, it works to query by role so wouldn't it be best to avoid `data-testid` or is this a different scenario? If it is, would it need to be mocked similarly to the following test since it isn't a child component? 

https://github.com/Automattic/wp-calypso/blob/4554587c1c81988ad5de50ee9b365892ee4f5951/client/signup/validation-fieldset/test/index.jsx#L32-L33

tyxla replied:
My understanding is that using `data-testid` in production code is not recommended, but when it's in mocks I find it just fine. Querying by role is just fine as well, though, so feel free to keep what you have.

brookewp replied:
Ah, okay - that makes sense! 👍

tyxla commented on client/components/marketing-survey/cancel-purchase-form/step-components/test/business-at-step.jsx:
```diff
@@ -1,75 +1,84 @@
 			);
-		} );
+		}
+		return null;
```
Should we return `content` in this case?

tyxla commented on client/components/marketing-survey/cancel-purchase-form/step-components/test/business-at-step.jsx:
```diff
@@ -1,75 +1,84 @@
-				<BusinessATStep translate={ translate } recordTracksEvent={ recordTracksEvent } />
+	const translate = ( content, params ) => {
+		if ( params && params.components ) {
+			return (
```
Should we still have `content` rendered somewhere in the mock?

tyxla replied:
This comment is still valid. If we have `components`, we'll never render the translated content in the mock, and I think we should.

brookewp replied:
The first set of tests under 'rendering translated content' test the translated content. Would it be redundant to render the translated content in the mock in the second set of tests? Or maybe it would make sense to combine them instead?

tyxla replied:
I think combining them would solve all the problems, yes 😉

tyxla reviewed (commented):
This generally looks good! 👍 

I've just left a couple of suggestions, let me know what you think.

brookewp committed:
Update tests based on PR feedback

brookewp commented:
Thanks again for your review, @tyxla! I believe I addressed most of your feedback, and I had one follow-up question. Let me know if I missed anything!

tyxla reviewed (approved):
Looks great 👍 Feel free to 🚢  as-is!
